docker/bash.sh

```diff
@@ -30,152 +33,356 @@
+
+DOCKER_FLAGS=( )
+DOCKER_ENV=( )
+DOCKER_MOUNT=( )
+DOCKER_DEVICES=( )
+
+
+# If the user gave a shortcut defined in the Jenkinsfile, use it.
+EXPANDED_SHORTCUT=$(lookup_image_spec "${DOCKER_IMAGE_NAME}")
+if [ -n "${EXPANDED_SHORTCUT}" ]; then
+    DOCKER_IMAGE_NAME="${EXPANDED_SHORTCUT}"
 fi
 
-DOCKER_IMAGE_NAME=$(lookup_image_spec "$1")
-if [ -z "${DOCKER_IMAGE_NAME}" ]; then
-    DOCKER_IMAGE_NAME=("$1")
+# Set up working directories
+
+DOCKER_FLAGS+=( --workdir "${REPO_MOUNT_POINT}" )
+DOCKER_MOUNT+=( --volume "${REPO_DIR}":"${REPO_MOUNT_POINT}"
+                --volume "${SCRIPT_DIR}":/docker
+              )
+
+# Set up CI-specific environment variables
+DOCKER_ENV+=( --env CI_BUILD_HOME="${REPO_MOUNT_POINT}"
+              --env CI_BUILD_USER="$(id -u -n)"
+              --env CI_BUILD_UID="$(id -u)"
+              --env CI_BUILD_GROUP="$(id -g -n)"
+              --env CI_BUILD_GID="$(id -g)"
+              --env CI_PYTEST_ADD_OPTIONS="${CI_PYTEST_ADD_OPTIONS:-}"
+              --env CI_IMAGE_NAME="${DOCKER_IMAGE_NAME}"
+            )
+
+
+# Pass tvm test data folder through to the docker container, to avoid
+# repeated downloads.  Check if we have permissions to write to the
+# directory first, since the CI may not.
+TEST_DATA_PATH="${TVM_DATA_ROOT_PATH:-${HOME}/.tvm_test_data}"
+if [[ -d "${TEST_DATA_PATH}" && -w "${TEST_DATA_PATH}" ]]; then
+    DOCKER_MOUNT+=( --volume "${TEST_DATA_PATH}":"${REPO_MOUNT_POINT}"/.tvm_test_data )
 fi
 
-if [ "$#" -eq 1 ]; then
-    COMMAND="bash"
-    interactive=1
+
+# Remove the container once it finishes running (--rm) and share the
+# PID namespace (--pid=host).  The process inside does not have pid 1
+# and SIGKILL is propagated to the process inside, allowing jenkins to
+# kill it if needed.
+DOCKER_FLAGS+=( --rm --pid=host)
+
+# Expose services running in container to the host.
+if $USE_NET_HOST; then
     if [[ $(uname) == "Darwin" ]]; then
         # Docker's host networking driver isn't supported on macOS.
         # Use default bridge network and expose port for jupyter notebook.
-        CI_DOCKER_EXTRA_PARAMS+=( "${CI_DOCKER_EXTRA_PARAMS[@]}" "-p 8888:8888" )
+        DOCKER_FLAGS+=( -p 8888:8888 )
     else
-        CI_DOCKER_EXTRA_PARAMS+=( "${CI_DOCKER_EXTRA_PARAMS[@]}" "--net=host" )
+        DOCKER_FLAGS+=(--net=host)
     fi
-else
-    shift 1
-    COMMAND=("$@")
 fi
 
-if [ $interactive -eq 1 ]; then
-    CI_DOCKER_EXTRA_PARAMS=( "${CI_DOCKER_EXTRA_PARAMS[@]}" -it )
+# Set up interactive sessions
+if ${INTERACTIVE}; then
+    DOCKER_FLAGS+=( --interactive --tty )
 fi
 
-SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
-WORKSPACE="$(pwd)"
-
-# Use nvidia-docker if the container is GPU.
-if [[ ! -z $CUDA_VISIBLE_DEVICES ]]; then
-    CUDA_ENV="-e CUDA_VISIBLE_DEVICES=${CUDA_VISIBLE_DEVICES}"
-else
-    CUDA_ENV=""
-fi
+# Expose external directories to the docker container
+for MOUNT_DIR in ${MOUNT_DIRS[@]+"${MOUNT_DIRS[@]}"}; do
```

Comment:
nit: extra } here and elsewhere.

i think the unbound variable was due to referencing an empty array. you need to modify all references to empty arrays to include the default substitution, so this should read like:

`for MOUNT_DIR in "${MOUNT_DIRS[@]-}"; do`

this preserves the quote-formatting to avoid shlexing array elements. could you update?

Reply:
I think the current implementation is the pedantically correct pattern, following [this stackoverflow post](https://stackoverflow.com/questions/7577052/bash-empty-array-expansion-with-set-u/61551944#61551944).  The main problem with the `"${MOUNT_DIRS[@]-}"` pattern is that it is vulnerable to typos, and `"${MOUNTT_DIRS[@]-}"` would silently be expanded to the empty string as well.  I tested the current version to verify that the array elements aren't shlexed apart, and the recommended variant does keep separate elements separate.

That said, I was going back and forth on it while writing it, because avoiding typos at the expense of making it nigh unreadable isn't really worth it.  Especially if we want to migrate this to a python script at some point, I'd go for version that's easier to read, and will change them to `"${MOUNT_DIRS[@]-}"`.

Reply:
And updated.

Reply:
And after some testing on my side, it looks like `"${MOUNT_DIRS[@]-}"` has the exact opposite problem.  It expands to an empty string, even when it should expand to an empty list.

* `"${ARR[@]}"` expands to the empty array on bash 5.x, but triggers `set -u` for empty arrays in bash 3.x.
* `"${ARR[@]-}"` avoids the `set -u` error, but expands to an empty string rather than an empty array.  This causes an error because docker interprets the empty string argument as the image name.
* `"${ARR[@]:-}"` has the same issue as `"${ARR[@]-}"`
* `${ARR[@]+"${ARR[@]}"}` works on my side, and avoids shlexing quoted inputs, but is absolutely ugly.

I'm reverting back to the `${ARR[@]+"${ARR[@]}"}`, but I would put this as further evidence to rewrite this script into python to avoid these issues.  @areusch Can you test on MacOSX + bash 3.2 to verify whether this syntax works there as well?

Reply:
alright thanks! i tested this locally and it seems to work fine actually. the mysteries of bash.